### PR TITLE
feat: implement concurrent export limit

### PR DESCRIPTION
**Pull request by javier-godoy**
## Summary by CodeRabbit

- **New Features**
  - Introduced concurrent download functionality for `GridExporter`, including timeout events and cost management.
  - Added a demo for handling large datasets with concurrent download settings.

- **Upgrades**
  - Upgraded `grid-exporter-addon` to version `2.4.0-SNAPSHOT`.
  - Updated Java compiler target and source versions to 17.

- **Bug Fixes**
  - Addressed concurrency issues in download management.

- **Tests**
  - Added extensive tests for concurrent download scenarios, including limits and timeouts.

- **Chores**
  - Updated `.gitignore` to exclude specific development and configuration files.

### CODE REVIEW

File: src/test/java/com/flowingcode/vaadin/addons/gridexporter/test/ConcurrentExportTests.java [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+  }
+
+  private class ConcurrentStreamResourceWriter
+      extends ConfigurableConcurrentStreamResourceWriter {
+
+    public ConcurrentStreamResourceWriter(StreamResourceWriter delegate) {
+      super(delegate);
+    }
+
+  }
```
[the remark this answers is not in the document]
Comment (javier-godoy) [reply src/test/java/com/flowingcode/vaadin/addons/gridexporter/test/ConcurrentExportTests.java:44]: @coderabbitai Yes, I wanted to "import" it with a different name ;)

File: src/main/java/com/flowingcode/vaadin/addons/gridexporter/GridExporter.java [hunk 2]
```diff
@@ -288,7 +305,105 @@ public StreamResource getExcelStreamResource() {
...
+   * 1.0 ({@link #DEFAULT_COST}) represents a standard unit of resource usage, while a cost of 0.5
+   * represents half a unit, and a cost above 1.0 indicates higher than normal resource usage.
+   * <p>
+   *
+   * If the cost is zero or negative, no permits are needed. However, any positive cost, no matter
+   * how small, will require at least one permit to prevent downloads with very low costs from
+   * bypassing the semaphore. {@link #MIN_COST} represents the minimal fractional cost that acquires
+   * only one permit (hence {@code 2*MIN_COST} acquires two permits and so on). A cost of
+   * {@link #MAX_COST} prevents any other downloads from acquiring permits simultaneously.
+   *
```
[the remark this answers is not in the document]
Comment (javier-godoy) [reply src/main/java/com/flowingcode/vaadin/addons/gridexporter/GridExporter.java:402]: @coderabbitai "If the cost is zero or negative, no permits are needed"

File: src/test/java/com/flowingcode/vaadin/addons/gridexporter/test/ConcurrentExportTests.java [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+    ConcurrentStreamResourceWriter.setLimit(0x10000);
+    Assert.assertEquals(GridExporter.MAX_COST, ConcurrentStreamResourceWriter.getLimit(), 0);
+
+    //Any positive cost, no matter how small, will require at least one permit.
+    ConcurrentStreamResourceWriter.setLimit(Float.MIN_NORMAL);
+    Assert.assertEquals(GridExporter.MIN_COST, ConcurrentStreamResourceWriter.getLimit(), 0);
+
+    ConcurrentStreamResourceWriter.setLimit(0.99f / 0x10000);
+    Assert.assertEquals(GridExporter.MIN_COST, ConcurrentStreamResourceWriter.getLimit(), 0);
+  }
```
[the remark this answers is not in the document]
Comment (javier-godoy) [reply src/test/java/com/flowingcode/vaadin/addons/gridexporter/test/ConcurrentExportTests.java:198]: @coderabbitai Please commit suggestion

**Comment by javier-godoy** [comment]
@coderabbitai full review

**Comment by javier-godoy** [comment]
@coderabbitai full review

**Comment by javier-godoy** [comment]
@coderabbitai Analyze ConcurrentStreamResourceWriter in detail

**Comment by javier-godoy** [comment]
@coderabbitai Explain how the number of permits is determined based on the calculated cost.

**Comment by javier-godoy** [comment]
@coderabbitai Describe in detail the algorithm implemented in `setLimit(float limit)` method of `ConcurrentStreamResourceWriter`

**Comment by javier-godoy** [comment]
@coderabbitai Describe in detail the algorithm implemented in `costToPermits(float cost, int maxPermits)` method of `ConcurrentStreamResourceWriter`

**Comment by javier-godoy** [comment]
@coderabbitai Explain how more robust exception handling could be implemented in ConcurrentStreamResourceWriter to manage scenarios where semaphore timeouts, interruptions or other failures occur.

**Review by mlopezFC — APPROVED** [review]
Excellent work, I do not have anything else to say besides what the 🐰 already said.

**Pull request merged**